### ISSUE TITLE
Clearing subscriptions throws SystemStackError

### ISSUE THREAD

**Issue by chadwilken** [rmosolgo/graphql-ruby#2701]
When trying to use the dashboard or `MySchema.subscriptions.clear` to reset all the subscriptions I am getting the following stacktrace.

```
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:43 :in `read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:397 :in `format_bulk_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:376 :in `format_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:365 :in `read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `block in format_multi_bulk_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `initialize`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `new`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `format_multi_bulk_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:377 :in `format_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:365 :in `read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `block in format_multi_bulk_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `initialize`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `new`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:406 :in `format_multi_bulk_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:377 :in `format_reply`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/connection/ruby.rb:365 :in `read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:265 :in `block in read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:253 :in `io`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:264 :in `read`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:123 :in `block in call`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:234 :in `block (2 levels) in process`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:375 :in `ensure_connected`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:224 :in `block in process`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:312 :in `logging`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:223 :in `process`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis/client.rb:123 :in `call`
[GEM_ROOT]/gems/scout_apm-2.6.6/lib/scout_apm/instruments/redis.rb:32 :in `block in call_with_scout_instruments`
[GEM_ROOT]/gems/scout_apm-2.6.6/lib/scout_apm/tracer.rb:34 :in `instrument`
[GEM_ROOT]/gems/scout_apm-2.6.6/lib/scout_apm/tracer.rb:44 :in `instrument`
[GEM_ROOT]/gems/scout_apm-2.6.6/lib/scout_apm/instruments/redis.rb:31 :in `call_with_scout_instruments`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis.rb:2617 :in `block in _scan`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis.rb:50 :in `block in synchronize`
/usr/local/rbenv/versions/2.5.7/lib/ruby/2.5.0/monitor.rb:235 :in `mon_synchronize`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis.rb:50 :in `synchronize`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis.rb:2616 :in `_scan`
[GEM_ROOT]/gems/redis-4.1.0/lib/redis.rb:2637 :in `scan`
[GEM_ROOT]/gems/graphql-pro-1.11.0/lib/graphql/pro/subscriptions/redis_storage.rb:97 :in `clear`
[GEM_ROOT]/gems/graphql-pro-1.11.0/lib/graphql/pro/subscriptions/redis_storage.rb:100 :in `clear`
[GEM_ROOT]/gems/graphql-pro-1.11.0/lib/graphql/pro/subscriptions/redis_storage.rb:100 :in `clear`
# This line happens like 900 more times…
```

**Comment by rmosolgo** [comment]
Yikes! Thanks for reporting this. 👀

**Comment by rmosolgo** [comment]
🤔 It seems like `.scan` isn't working quite the way I thought it would. 

Does this error surface in development for you? If so, could you add this monkey-patch, then try again to trigger the error?

```ruby 
# config/initializers/graphql_dashboard_debug.rb
module GraphQL
  module Pro
    class Subscriptions < GraphQL::Subscriptions
      class RedisStorage
        def clear(cursor: 0)
          next_cursor, keys = @redis.scan(cursor, match: "#{PREFIX}*")
          # Print the result from `SCAN`
          p [next_cursor, keys]
          keys.each { |k| @redis.del(k) }
          if next_cursor != "0"
            clear(cursor: next_cursor)
          end
        end
      end 
    end 
  end 
end 
```

Could you share the output when you run it? I _expect_ `next_cursor` to be a string, [where `"0"` indicates a full iteration](https://redis.io/commands/scan#scan-basic-usage), but I must be missing something!

**Comment by chadwilken** [comment]
I haven't been able to repro in dev, granted I haven't tried to use a ton of subscriptions. In production, where this is breaking, it was attempting to clear ~40k topics worth of subscriptions. The solution that I ran to clear subscriptions was:
```
keys = redis.keys("#{PATTERN}*")
redis.del(*keys) if keys.present?
```
Granted that is probably memory intensive so not the best way to use in the lib.

I wonder if bumping the `count` to something larger so that there is a lower chance of `match` returning zero elements would help. It seems that Ruby's stack limit isn't great for a ton of recursion. Idk what the tradeoffs are if you bumped the count to something larger like 1000 with the understanding that you probably won't get that many. I think a `loop` might get around this recursion issue since it would be iterating rather than recursing a method. Something like:
```
def clear
  next_cursor, keys = @redis.scan(0, match: "#{PREFIX}*")

  loop do
    if keys.present?
      @redis.del(*keys)
    end

    break if next_cursor == "0"

    next_cursor, keys = @redis.scan(next_cursor, match: "#{PREFIX}*")
  end
end
```

**Comment by rmosolgo** [comment]
I just released graphql-pro 1.13.0 with a refactor to this method like your suggestion, thanks! In fact, I only imagined that it was an infinite loop ... I didn't consider the legitimate possibility that there were just too many calls to the recursive method! 

🤞 Maybe someday we'll get tail-call optimization by default in Ruby :P
